src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/AccessRightService.java

```diff
@@ -313,108 +381,161 @@ private ServiceClientDto accessRightTypeToServiceClientDto(AccessRightType acces
-            txSubjects.addAll(txLocalGroupXroadIds);
         }
-        return txSubjects;
-    }
 
-    /**
-     * Verify that all identifiers are authentic, then get the existing ones from the local db and persist
-     * the not-existing ones. This is a necessary step if we are changing identifier relations (such as adding
-     * access rights to services)
-     * @param subsystemIds {@link GlobalGroupId} or {@link ClientId}
-     * @return List of XRoadIds ({@link GlobalGroupId} or {@link ClientId})
-     */
-    private Set<XRoadId> getOrPersistSubsystemIds(Set<XRoadId> subsystemIds)
-            throws IdentifierNotFoundException {
-        // Check that the identifiers exist in globalconf
-        // LocalGroups must be verified separately! (they do not exist in globalconf)
-        if (!globalConfService.clientIdentifiersExist(subsystemIds)) {
-            // This exception should be pretty rare since it only occurs if bogus subjects are found
-            throw new IdentifierNotFoundException();
+        // list endpoints, which this subject / service client has already been granted access to
+        Set<EndpointType> existingAccessibleEndpoints = clientType.getAcl().stream()
+                .filter(accessRightType -> accessRightType.getSubjectId().equals(subjectId))
+                .map(accessRightType -> accessRightType.getEndpoint())
+                .collect(Collectors.toSet());
+
+        if (existingAccessibleEndpoints.contains(endpoint)) {
+            throw new DuplicateAccessRightException("Subject " + subjectId.toShortString()
+                    + " already has an access right for endpoint " + endpoint.getId());
         }
-        return identifierService.getOrPersistXroadIds(subsystemIds);
+
+        AccessRightType newAccessRight = new AccessRightType();
+        newAccessRight.setEndpoint(endpoint);
+        newAccessRight.setSubjectId(subjectId);
+        newAccessRight.setRightsGiven(rightsGiven);
+        clientType.getAcl().add(newAccessRight);
+
+        // return a dto
+        ServiceClientAccessRightDto dto = ServiceClientAccessRightDto.builder()
+                .serviceCode(endpoint.getServiceCode())
+                .rightsGiven(FormatUtils.fromDateToOffsetDateTime(rightsGiven))
+                .title(getServiceTitle(clientType, endpoint.getServiceCode()))
+                .build();
+        return dto;
+    }
+
+    // TO DO: currently duplicate with ServiceClientService, not sure if ServiceClientService will be refactored,
+    // and what is the correct place if both need this
+    private String getServiceTitle(ClientType clientType, String serviceCode) {
+        ServiceType service = clientType.getServiceDescription().stream()
+                .flatMap(sd -> sd.getService().stream())
+                .filter(serviceType -> serviceType.getServiceCode().equals(serviceCode))
+                .findFirst()
+                .get();
+
+        return service == null ? null : service.getTitle();
```

Comment:
I think this was a real bug? Good catch. 

Fixed in https://github.com/nordic-institute/X-Road/commit/860f81a6275ccf04748ab75c4e76d6e48e1a4fa7 in branch XRDDEV-962-remove-service-client-access-rights since getServiceTitle has been refactored to a different class and it was easier to fix like that.